graylog2-server/src/main/java/org/graylog2/bindings/GraylogNodeModule.java

```diff
@@ -0,0 +1,95 @@
+ */
+package org.graylog2.bindings;
+
+import com.google.common.eventbus.EventBus;
+import com.google.inject.Scopes;
+import org.glassfish.jersey.internal.guava.Sets;
+import org.graylog2.GraylogNodeConfiguration;
+import org.graylog2.audit.AuditBindings;
+import org.graylog2.bindings.providers.ClusterEventBusProvider;
+import org.graylog2.configuration.EventBusConfiguration;
+import org.graylog2.configuration.MongoDbConfiguration;
+import org.graylog2.events.ClusterEventBus;
+import org.graylog2.jackson.InputConfigurationBeanDeserializerModifier;
+import org.graylog2.plugin.LocalMetricRegistry;
+import org.graylog2.plugin.inject.Graylog2Module;
+import org.graylog2.plugin.system.FilePersistedNodeIdProvider;
+import org.graylog2.plugin.system.NodeId;
+import org.graylog2.plugin.system.SimpleNodeId;
+import org.graylog2.security.encryption.EncryptedValueService;
+import org.graylog2.shared.bindings.ObjectMapperModule;
+import org.graylog2.shared.bindings.SchedulerBindings;
+import org.graylog2.shared.bindings.ServerStatusBindings;
+import org.graylog2.shared.bindings.providers.EventBusProvider;
+
+import java.util.Set;
+
+/**
+ * <p>Guice module that contains all necessary bindings to start a basic node in a Graylog cluster.</p>
+ * <p>Bindings will be loaded conditionally based on the configuration in {@link GraylogNodeConfiguration}</p>
+ */
+public class GraylogNodeModule extends Graylog2Module {
+    private final GraylogNodeConfiguration configuration;
+
+    public GraylogNodeModule(final GraylogNodeConfiguration configuration) {
+        this.configuration = configuration;
+    }
+
+    @Override
+    protected void configure() {
+        bind(GraylogNodeConfiguration.class).toInstance(configuration);
+        if (configuration.withMongoDb()) {
+            install(new MongoDbConnectionModule());
+            install(new ObjectMapperModule());
+        }
+        if (configuration.withScheduler()) {
+            install(new SchedulerBindings());
+        }
+        install(new AuditBindings());
+
+        if (configuration.withEventBus()) {
+            bind(ClusterEventBus.class).toProvider(ClusterEventBusProvider.class).asEagerSingleton();
+            bind(EventBus.class).toProvider(EventBusProvider.class).in(Scopes.SINGLETON);
+        }
+        // ensure we always create a new LocalMetricRegistry, they are meant to be separate from each other
+        bind(LocalMetricRegistry.class).in(Scopes.NO_SCOPE);
+
+        if (configuration.withNodeIdFile()) {
+            bind(NodeId.class).toProvider(FilePersistedNodeIdProvider.class).asEagerSingleton();
+        } else {
+            bind(NodeId.class).toInstance(new SimpleNodeId("dummy-nodeid"));
+        }
+
+        install(new ServerStatusBindings(configuration.withCapabilities()));
+
+        bind(EncryptedValueService.class).asEagerSingleton();
+        bind(InputConfigurationBeanDeserializerModifier.class).toInstance(InputConfigurationBeanDeserializerModifier.withoutConfig());
```

Comment:
@moesterheld: For the server, this binding breaks the saving of inputs that contain encrypted values. You can reproduce it by creating an `AWS CloudTrail` where you only set the title (and node) and leave everything else empty. You can save it, but afterwards the inputs page is broken and you have to remove the input manually from MongoDB.

We should either go back to a just-in-time binding (for the server) or do a regular singleton scope binding. I didn't test the latter but it should be fine.

Reply:
Thanks @thll , I will fix this and try to add an automated test that covers this somehow.

Reply:
Thank you! Yeah, having some kind of integration test for this would be great as the cause is very subtle and hard to find.